### PR TITLE
initial step version

### PR DESCRIPTION
**Pull request by trapacska**

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[[constraint]]
+  name = "github.com/bitrise-io/bitrise"
+  version = "1.27.1"
```
Comment (godrei) [inline Gopkg.toml]: pls lock for master branches

File: README.md [hunk 2]
```diff
@@ -1 +1,54 @@
-# bitrise-step-flank
\ No newline at end of file
+# Flank
+
+> Run your tests using Flank. The step will automaticall detect which project type your flank config uses and the corresponding flank command fille be run.
```
Comment (godrei) [inline README.md]: possible typo: `fille` ?

File: Gopkg.toml (hunk 1 above)
Comment (godrei) [inline Gopkg.toml]: pls lock for master branches

File: main.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+		failf("Failed to split command flags, error: %s", err)
+	}
+
+	fmt.Println()
+	command := command.New("java", append([]string{"-jar", binaryPath, platform, "run", "-c", cfg.ConfigPath}, commandFlags...)...).
+		SetStdin(os.Stdin).
+		SetStdout(os.Stdout).
+		SetStderr(os.Stderr)
+
+	log.Donef("$ %s", command.PrintableCommandArgs())
```
Comment (lszucs) [inline main.go:266]: Logging done before actually running the command is confusing when debugging.

File: README.md [hunk 4]
```diff
@@ -1 +1,54 @@
...
+    > ```yaml
+    > format_version: "6"
+    > default_step_lib_source: https://github.com/bitrise-io/bitrise-steplib.git
+    > 
+    > workflows:
+    >   my-workflow:
+    >     steps:
+    >     - path::./:
+    >         inputs: 
+    >         - my_input: "my input value"
```
Comment (godrei) [inline README.md]: pls list the required inputs here

File: bitrise.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
+format_version: 4
```
Comment (godrei) [inline bitrise.yml]: pls use the latest format version

File: bitrise.yml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            set -ex
+
+            echo "$FLANK_CONFIG" > ./flank.yml
+    - path::./:
+        title: Test Step
+        inputs:
+        - google_service_account_json: $GOOGLE_SERVICE_ACCOUNT
+        - config_path: ./flank.yml
+
+  audit-this-step:
```
Comment (godrei) [inline bitrise.yml:100]: pls wire it in for the test wf

File: bitrise.yml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    steps:
+    - go-list:
+    - golint:
+    - errcheck:
+    - go-test:
+    - script:
+        title: "clean-up"
+        inputs:
+        - content: |-
+            #!/bin/bash
```
Comment (godrei) [inline bitrise.yml]: pls use `#!/usr/bin/env bash`

File: bitrise.yml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+            rm -rf ./_tmp
+    - change-workdir:
+        title: Switch working dir to test/_tmp dir
+        run_if: true
+        inputs:
+        - path: ./_tmp
+        - is_create_path: true
+    - script:
+        title: "Git clone sample android project"
```
Comment (godrei) [inline bitrise.yml:37]: pls be consistent with titles, either use it for every step or not. Either start with upper case letter or not.

**Review by lszucs — COMMENTED** [review]
Left a comment.

File: main.go [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+	"path/filepath"
+	"strings"
+	"time"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/bitrise-io/go-utils/pathutil"
+
+	"github.com/hashicorp/go-version"
+
```
Comment (godrei) [inline main.go]: pls remove empty newlines in import list

File: main.go [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+	// eae4b1e50e7eb05c02fc678994c6425bf14e1701	refs/tags/v2.0.0
+
+	var versions []string
+	for _, line := range strings.Split(out, "\n") {
+		if ref := strings.Split(line, "\t"); len(ref) == 2 {
+			versions = append(versions, strings.TrimPrefix(ref[1], "refs/tags/"))
+		}
+	}
+
+	lastVersion, err := version.NewVersion("0.0.0")
```
Comment (godrei) [inline main.go:82]: is `0.0.0` an existing version, what if no other version would found?

File: main.go [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+	if err := os.MkdirAll(filepath.Dir(pth), 0777); err != nil {
+		return err
+	}
+	return fileutil.WriteStringToFile(pth, cred)
+}
+
+// gets the tags list, splits the lines per tab and finds the prefix-truncated version strings
+// if the version string is a valid semver version then this function returns the latest one
+func getLatestVersion() (string, error) {
+	cmd := command.New("git", "ls-remote", "--tags", "--quiet", "https://github.com/TestArmada/flank")
```
Comment (godrei) [inline main.go]: `baseURL` holds `https://github.com/TestArmada/flank`

File: main.go [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+	out, err := cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return "", err
```
Comment (godrei) [inline main.go]: if exit status error present return out

File: main.go [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+	}
+
+	// hashes and version refs are separated by tab, output will be something like:
+	// a85cdf850b7562d38559b84d405c97f7d194603s refs/tags/flank_snapshot
+	// f0d902fdd947e7f2db43213929bdc6069a681624	refs/tags/v1.4.2
+	// 4f00292267eaa2b230c4a5705a50c5302d7f9dc1	refs/tags/v1.5.0
```
Comment (godrei) [inline main.go]: 1 comment is enough for 1 example type

**Review by lpusok — COMMENTED** [review]
👍

File: main.go [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+			}
+		}
+	}
+
+	return lastVersion.String(), nil
+}
+
+// if input version is latest then it returns the fetched latest release version download url otherwise
+// returns the release version download url for the given version
+func getDownloadURLbyVersion(version string) (string, error) {
```
Comment (godrei) [inline main.go]: could be shorter

```
func getDownloadURLbyVersion(version string) (string, error) {
	if version == "latest" {
		var err error
		if version, err = getLatestVersion(); err != nil {
			return "", nil
		}
	}
	return fmt.Sprintf("%s/releases/download/v%s/flank.jar", baseURL, version), nil
}
```

File: main.go [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+	}
+
+	var latestDir string
+	var latestModtime time.Time
+
+	for _, fInf := range fInfs {
+		if !fInf.IsDir() {
+			continue
+		}
+		if fInf.ModTime().After(latestModtime) || fInf.ModTime().Equal(latestModtime) {
```
Comment (godrei) [inline main.go]: `if !fInf.ModTime().Before(latestModtime) {`

File: main.go [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+
+		srcFile := filepath.Join(latestDir, fInf.Name())
+		destinationFile := filepath.Join(os.Getenv("BITRISE_DEPLOY_DIR"), fInf.Name())
+
+		data, err := ioutil.ReadFile(srcFile)
+		if err != nil {
+			return err
+		}
+		err = ioutil.WriteFile(destinationFile, data, 0644)
+		if err != nil {
```
Comment (godrei) [inline main.go]: `if err := ioutil.WriteFile(destinationFile, data, 0644); err != nil {`

File: step.yml [hunk 17]
```diff
@@ -0,0 +1,46 @@
+title: Flank
```
Comment (godrei) [inline step.yml:1]: What about: Flank test?

File: step.yml [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+summary: "Run your tests using Flank."
+description: "Run your tests using Flank. The step will automaticall detect which project type your flank config uses and the corresponding flank command fille be run."
```
Comment (godrei) [inline step.yml]: The step will __automatically__ detect which project type your flank config uses and the corresponding flank command __will__ be run.

File: step.yml [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+    opts:
+      title: "Version"
+      summary: "Flank binary version."
+      description: "Flank binary version. You can use any tag name that is available on https://github.com/TestArmada/flank/releases or latest which will download the latest non-pre-elease version."
+      is_required: true
+  - command_flags:
+    opts:
+      title: "Command Flags"
+      summary: "These flags will be appended to the flank command."
+      description: "These flags will be appended to the flank command."
```
Comment (godrei) [inline step.yml]: pls mention the generated commands

File: main.go [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+	pth := fmt.Sprintf(credentialPathFmt, os.Getenv("HOME"))
+	if err := os.MkdirAll(filepath.Dir(pth), 0777); err != nil {
+		return err
+	}
+	return fileutil.WriteStringToFile(pth, cred)
+}
+
+// gets the tags list, splits the lines per tab and finds the prefix-truncated version strings
+// if the version string is a valid semver version then this function returns the latest one
+func getLatestVersion() (string, error) {
```
Comment (godrei) [inline main.go]: write test for the git command output parsing

File: main.go [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+
+type config struct {
+	ServiceAccountJSON stepconf.Secret `env:"google_service_account_json,required"`
+	ConfigPath         string          `env:"config_path,file"`
+	Version            string          `env:"version,required"`
+	CommandFlags       string          `env:"command_flags"`
+}
+
+// returns android if there is an app field under gcloud in the config yml
+func detectPlatform(configYMLPath string) (string, error) {
```
Comment (godrei) [inline main.go:37]: write test for it

File: main.go (hunk 14 above)
Comment (godrei) [inline main.go]: write test

File: main.go [hunk 22]
```diff
@@ -0,0 +1,280 @@
...
+}
+
+func failf(format string, args ...interface{}) {
+	log.Errorf(format, args...)
+	os.Exit(1)
+}
+
+// lists all dirs inside of ./results dir and selects the latest(by modtime)
+// then all the files in the root will be copied from this dir to the root of the dir under BITRISE_DEPLOY_DIR
+func exportArtifacts(copied func(src, dest string)) error {
```
Comment (godrei) [inline main.go]: write test

**Review by godrei — CHANGES_REQUESTED** [review]
Pls have a look at my notes

File: main.go (hunk 3 above)
Comment (trapacska) [reply main.go:266]: We usually print commands that will be run with green and starting with a `$`.

File: main.go (hunk 14 above)
Comment (trapacska) [reply main.go]: There is an error in the code sample at return "", nil in line 5. Error should be error and not nil. This was an issue in the original code as well.

File: step.yml (hunk 17 above)
Comment (trapacska) [reply step.yml:1]: The step name is already reviewed and discussed by the proposal. Flank test for me means a new type of test and not the test runner what Flank really is.

File: bitrise.yml [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+default_step_lib_source: https://github.com/bitrise-io/bitrise-steplib.git
+
+app:
+  envs:
+  - GOOGLE_SERVICE_ACCOUNT: $GOOGLE_SERVICE_ACCOUNT
+  - FLANK_CONFIG: $FLANK_CONFIG
+  - SAMPLE_ANDROID_APP_GIT_URL: "https://github.com/bitrise-samples/android-multiple-test-results-sample"
+
+workflows:
+  test:
```
Comment (BirmacherAkos) [inline bitrise.yml:14]: Could you wire in an iOS test as well? If the step supports iOS too [(iOS support in step.yml)](https://github.com/bitrise-steplib/bitrise-step-flank/pull/1/files#diff-91d8d269a1c9bd3631ddb04b80b01c4eR12) , then I think it should test it too.

**Review by BirmacherAkos — CHANGES_REQUESTED** [review]
Only one request/question. Other than that it looks good to me 👍

File: main.go [hunk 24]
```diff
@@ -63,40 +58,30 @@ func detectPlatform(configYMLPath string) (string, error) {
...
-func getLatestVersion() (string, error) {
-	cmd := command.New("git", "ls-remote", "--tags", "--quiet", "https://github.com/TestArmada/flank")
-	out, err := cmd.RunAndReturnTrimmedCombinedOutput()
-	if err != nil {
-		return "", err
+	pth := filepath.Join(tmpPth, "cred.json")
+	if err := fileutil.WriteStringToFile(pth, cred); err != nil {
+		return err
 	}
+	return os.Setenv("GOOGLE_APPLICATION_CREDENTIALS", pth)
```
Comment (godrei) [inline main.go:69]: return the path and expose this env only for the command which uses it

File: main.go [hunk 25]
```diff
@@ -63,40 +58,30 @@ func detectPlatform(configYMLPath string) (string, error) {
 
 // stores string under the default gcloud credential's local path
 func storeCredentials(cred string) error {
```
Comment (godrei) [inline main.go:60]: update the comment

**Review by lszucs — COMMENTED** [review]
LGTM

File: main.go [hunk 26]
```diff
@@ -63,40 +58,30 @@ func detectPlatform(configYMLPath string) (string, error) {
...
+func parseGitTags(gitOutput string) (versions []string) {
+	for _, line := range strings.Split(gitOutput, "\n") {
 		if ref := strings.Split(line, "\t"); len(ref) == 2 {
 			versions = append(versions, strings.TrimPrefix(ref[1], "refs/tags/"))
 		}
 	}
+	return
+}
 
+func findLatestVersion(versions []string) string {
```
Comment (godrei) [inline main.go:81]: i would return the version object (nil if no version found)

**Review by godrei — APPROVED** [review]
pls have a llok at my note

**Review by lpusok — COMMENTED** [review]
lgtm

File: main.go (hunk 10 above)
Comment (lpusok) [reply main.go:82]: It is not valid semver version tough.

**Review by BirmacherAkos — APPROVED** [review]

**Review by lszucs — COMMENTED** [review]
LGTM

**Review by godrei — APPROVED** [review]

**Review by lpusok — COMMENTED** [review]
👍

**Pull request merged**